source/common/router/config_impl.cc

```diff
@@ -79,6 +83,15 @@ bool RouteEntryImplBase::matches(const Http::HeaderMap& headers, uint64_t random
     matches &= (headers.get(Http::Headers::get().ContentType) == content_type_);
   }
 
+  if (!config_headers_.empty()) {
+    for (const Json::Object& header_map : config_headers_) {
```

Comment:
That is a breaking change. We can look into that separately (open internal JIRA), but let's not do that as part of this change.